### PR TITLE
koordlet: change metric cache to tsdb

### PR DESCRIPTION
**Pull request by zwzhang0107**
### Ⅰ. Describe what this PR does

`metric cache` is a module for metric data persistence in koordlet.
Currently, it use `sqlite3` as the storage, which is an structured SQL-like database.
`sqlite3` shows poor performance for time-series data like resource metrics, consuming more space and slow on query.
Besides, as an structured database, tables and interfaces must be added when adding new metrics on `metric cache`, which shows bad expansibility。
We plan to replace `sqlite3` with `tsdb` for better performance and extension.

This PR finish two jobs:
- Finish the tsdb related framework in metric cache, providing append and query interface
- Replace `cpu throttled metric` storage with tsdb and adapt the QoS plugin `cpu_burst` and Collector `pod throttled` with new interface.

### Ⅱ. Does this pull request fix one issue?

see #586

### Ⅲ. Describe how to verify it

Resource Consumption Comparation: before & after pod & container throttled ratio saved to tsdb.

We deploy koordlet v1.2 and latest version with tsdb refactor on the same node and start 60 pods on it for data collecting.

In v1.2, only latest 30min metrics are saved in sqlite, in latest version the time range is extended to 24h by default.

image

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/metriccache/tsdb_storage.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+// TSDBStorage defines time-series type DB, providing insert and query interface
+type TSDBStorage interface {
+	Appendable
+	Queryable
+
+	// Close closes the storage and all its underlying resources.
+	Close() error
+}
+
```
Comment (hormes) [inline pkg/koordlet/metriccache/tsdb_storage.go:39]: how to truncate the old data intsdb storage?

Comment (zwzhang0107) [reply pkg/koordlet/metriccache/tsdb_storage.go:39]: https://github.com/koordinator-sh/koordinator/pull/1228/files#diff-45a081cdd0ea16350b2cdeb04da7a47f747a809739ef2c1ece292b4e1ec9a89dL26-L31

RetentionDuration and MaxBytes options are supported by prometheus.tsdb.

File: pkg/koordlet/metriccache/metric_types.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+	NodeMetricMemoryUsage  MetricKind = "node_memory_usage"
+	NodeMetricGPUCoreUsage MetricKind = "node_gpu_core_usage"
+	NodeMetricGPUMemUsage  MetricKind = "node_gpu_memory_usage"
+	NodeMetricGPUMemTotal  MetricKind = "node_gpu_memory_total"
+
+	PriorityMetricCPUUsage     MetricKind = "priority_cpu_usage"
+	PriorityMetricCPURealLimit MetricKind = "priority_cpu_real_limit"
+	PriorityMetricCPURequest   MetricKind = "priority_cpu_request"
+
+	PocMetricCPUUsage     MetricKind = "pod_cpu_usage"
```
Comment (LambdaHJ) [inline pkg/koordlet/metriccache/metric_types.go:43]: PodMetricMemoryUsage

File: pkg/koordlet/metriccache/metric_types.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+	PocMetricMemoryUsage  MetricKind = "pod_memory_usage"
```
Comment (LambdaHJ) [inline pkg/koordlet/metriccache/metric_types.go:44]: PodMetricMemoryUsage

File: pkg/koordlet/metriccache/metric_types.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+	PodMetricGPUMemUsage  MetricKind = "pod_gpu_memory_usage"
+	// PodMetricGPUMemTotal       MetricKind = "pod_gpu_memory_total"
+
+	ContainerMetricCPUUsage     MetricKind = "container_cpu_usage"
+	ContainerMetricMemoryUsage  MetricKind = "container_memory_usage"
+	ContainerMetricGPUCoreUsage MetricKind = "container_gpu_core_usage"
+	ContainerMetricGPUMemUsage  MetricKind = "container_gpu_memory_usage"
+	// ContainerMetricGPUMemTotal       MetricKind = "container_gpu_memory_total"
+
+	PocMetricCPUThrottled       MetricKind = "pod_cpu_throttled"
```
Comment (LambdaHJ) [inline pkg/koordlet/metriccache/metric_types.go:55]: ditto

File: pkg/koordlet/metriccache/metric_resources.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+package metriccache
+
+var (
+	defaultMetricFactory = NewMetricFactory()
+
+	// define all kinds of MetricResource
+	NodeCPUUsageMetric          = defaultMetricFactory.New(NodeMetricCPUUsage)
+	NodeMemoryUsageMetric       = defaultMetricFactory.New(NodeMetricMemoryUsage)
+	PodCPUUsageMetric           = defaultMetricFactory.New(PocMetricCPUUsage).withPropertySchema(MetricPropertyPodUID)
+	PodCPUThrottledMetric       = defaultMetricFactory.New(PocMetricCPUThrottled).withPropertySchema(MetricPropertyPodUID)
```
Comment (LambdaHJ) [inline pkg/koordlet/metriccache/metric_resources.go:26]: typora PocMetricCPUUsage

**Review by LambdaHJ — APPROVED** [review]
/lgtm

File: config/manager/koordlet.yaml [hunk 6]
```diff
@@ -130,3 +132,7 @@ spec:
             path: /dev
             type: ""
           name: host-dev
+        - emptyDir:
+            medium: Memory
```
Comment (hormes) [inline config/manager/koordlet.yaml:136]: Should we default to memory usage, the cost is a bit high?

**Review by hormes — APPROVED** [review]

**Pull request merged**